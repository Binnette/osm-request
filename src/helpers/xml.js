import { parse as xmlParse, serialize as xmlSerialize } from 'simple-xml-dom';
import packageJson from '../../package.json';

const osmRequestVersion = packageJson.version;

/**
 * Build a stringified OSM changeset
 * @param {string} [createdBy]
 * @param {string} [comment]
 * @return {string}
 */
export function buildChangesetXml(createdBy = '', comment = '') {
  const xmlString = `
    <osm>
      <changeset>
        <tag k="created_by" v="${createdBy}"/>
        <tag k="created_by:library" v="OSM Request ${osmRequestVersion}"/>
        <tag k="comment" v="${comment}"/>
      </changeset>
    </osm>
  `;

  return xmlSerialize(xmlParse(xmlString));
}

/**
 * Convert an XML into a JSON object
 * @param  {Object} xml
 * @return {Object}
 */
export function xmlToJson(xml) {
  let obj = {};

  if (xml.nodeType === 1) {
    for (let j = 0; j < xml.attributes.length; j++) {
      const attribute = xml.attributes.item(j);
      obj[attribute.nodeName] = attribute.nodeValue;
    }
  } else if (xml.nodeType === 3) {
    // text
    obj = xml.nodeValue.trim();
  }

  // do children
  if (xml.hasChildNodes()) {
    for (let i = 0; i < xml.childNodes.length; i++) {
      const item = xml.childNodes.item(i);
      const nodeName = item.nodeName;

      if (typeof obj[nodeName] === 'undefined') {
        const tmp = xmlToJson(item);

        if (tmp !== '') {
          if (tmp['#text']) {
            obj[nodeName] = tmp['#text'];
          } else {
            obj[nodeName] = tmp;
          }
        }
      } else {
        if (typeof obj[nodeName].push === 'undefined') {
          const old = obj[nodeName];
          obj[nodeName] = [old];
        }

        const tmp = xmlToJson(item);

        if (tmp !== '') {
          obj[nodeName].push(tmp);
        }
      }
    }
  }

  return obj;
}
