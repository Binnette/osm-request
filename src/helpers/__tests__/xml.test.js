import { buildChangesetXml } from '../xml';

jest.mock('../../../package.json', () => ({
  version: '1.2.3'
}));

describe('XML helpers', () => {
  describe('buildChangesetXml', () => {
    it('Should build a stringified OSM changeset', () => {
      expect(buildChangesetXml('me', 'my comment')).toMatchSnapshot();
      expect(buildChangesetXml()).toMatchSnapshot();
    });
  });
});
