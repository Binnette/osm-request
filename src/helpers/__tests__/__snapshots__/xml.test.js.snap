// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XML helpers buildChangesetXml Should build a stringified OSM changeset 1`] = `
"
    <osm>
      <changeset>
        <tag k=\\"created_by\\" v=\\"me\\"/>
        <tag k=\\"created_by:library\\" v=\\"OSM Request 1.2.3\\"/>
        <tag k=\\"comment\\" v=\\"my comment\\"/>
      </changeset>
    </osm>"
`;

exports[`XML helpers buildChangesetXml Should build a stringified OSM changeset 2`] = `
"
    <osm>
      <changeset>
        <tag k=\\"created_by\\" v=\\"\\"/>
        <tag k=\\"created_by:library\\" v=\\"OSM Request 1.2.3\\"/>
        <tag k=\\"comment\\" v=\\"\\"/>
      </changeset>
    </osm>"
`;
